_posts/2020-07-21-Securing-Applications-Azure-AD.md

```diff
@@ -0,0 +1,405 @@
+```
+
+In a real production configuration, I personally prefer to use the issuer URI as it offers most configuration via a single configuration property.
+This will issue a network call to the discovery document when the application starts, so when testing in an environment where Azure AD is not reachable, this will cause the application to crash.
+This is where the JWK URI can save the day.
+
+# The Angular Part
+
+When we now browse to any back-end endpoint, we receive: HTTP 401 Unauthorized.  
+Let's fix this in our [example Angular application](https://github.com/jmeys/azure-ad-demo-frontend){:target="_blank" rel="noopener noreferrer"}.
+
+## The Angular Library
+
+Azure AD has quickstart guides for different kinds of applications. 
+For Angular, however, the [msal-angular library](https://www.npmjs.com/package/@azure/msal-angular){:target="_blank" rel="noopener noreferrer"} currently only supports the implicit flow.
+Since the current best practices draft strongly discourages the implicit flow in favour of the authorization code flow with PKCE, we will look for an alternative.
+
+> The Microsoft Authentication Library for JavaScript (MSAL) should have support for PKCE soon, but at the time of writing, this feature was still in alpha.
+
+Even when there will be Microsoft libraries that can solve this problem, I try to stay vendor-neutral whenever possible.
+This makes it relatively easy to switch from one OAuth provider to another one like [Auth0](https://auth0.com/){:target="_blank" rel="noopener noreferrer"}, [AWS Cognito](https://aws.amazon.com/cognito/){:target="_blank" rel="noopener noreferrer"}, [Okta](https://www.okta.com/){:target="_blank" rel="noopener noreferrer"}, [Keycloak](https://www.keycloak.org/){:target="_blank" rel="noopener noreferrer"}, ...
+The only downside is that vendor-specific features will not be available.
+An example of this is the [On-Behalf-Of flow (OBO)](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow){:target="_blank" rel="noopener noreferrer"}, which is only supported by the Microsoft libraries.
+
+Since OAuth and OIDC are standards, we should be able to use any (certified) library which supports these.
+I say "should", as the specifications left a lot of room for tinkering and additions. 
+We'll see what I mean by this during the implementation.
+
+My favourite go-to library is [angular-oauth2-oidc](https://github.com/manfredsteyer/angular-oauth2-oidc){:target="_blank" rel="noopener noreferrer"} by Manfred Steyer. 
+This is also the one we'll use in this example.
+
+> Alternatively, you can use the [msal-angular library](https://www.npmjs.com/package/@azure/msal-angular){:target="_blank" rel="noopener noreferrer"} if you are fine with the implicit flow for now.
+
+## Example Application: Tour of Heroes
+
+As an example of an Angular application, we can use the [Tour of Heroes](https://angular.io/tutorial) Angular tutorial application. 
```

Comment:
```suggestion
As an example of an Angular application, we can use the [Tour of Heroes](https://angular.io/tutorial){:target="_blank" rel="noopener noreferrer"} Angular tutorial application. 
```

Reply:
So, is the FE application specified above based on the Tour of Heroes application of is it something different? This because you mentioned that we will integrate security to that FE application but here you tell the reader that he/she has to start with the Angular tutorial application.

Maybe just need to clear that out.